lib/pos_blockchain.go

```diff
@@ -100,160 +101,165 @@ func (bc *Blockchain) processBlockPoS(block *MsgDeSoBlock, currentView uint64, v
-// case, we will add the block to the block index with status STORED and return the BlockNode.
-func (bc *Blockchain) isValidPoSBlock(block *MsgDeSoBlock) (*BlockNode, error) {
-	var storeInBlockIndexErr error
+//   - Base case: If block is VALIDATED or VALIDATE_FAILED, return the BlockNode as-is.
+//   - Recursive case: If the block is not VALIDATED or VALIDATE_FAILED in the blockIndex, we will perform all
+//     validations and add the block to the block index with the appropriate status (VALIDATED OR VALIDATE_FAILED) and
+//     return the new BlockNode.
+//   - Error case: Something goes wrong that doesn't result in the block being marked VALIDATE or VALIDATE_FAILED. In
+//     this case, we will add the block to the block index with status STORED and return the BlockNode.
+func (bc *Blockchain) validateAndIndexBlockPoS(block *MsgDeSoBlock) (*BlockNode, error) {
 	blockHash, err := block.Header.Hash()
 	if err != nil {
-		return nil, errors.Wrapf(err, "isValidPoSBlock: Problem hashing block %v", block)
+		return nil, errors.Wrapf(err, "validateAndIndexBlockPoS: Problem hashing block %v", block)
 	}
-	// Base case - Check if the block itself is validated or validate failed. If so, we can return early.
+
+	// Base case - Check if the block is validated or validate failed. If so, we can return early.
 	blockNode, exists := bc.blockIndex[*blockHash]
-	if exists {
-		if blockNode.IsValidateFailed() {
-			return blockNode, errors.New("isValidPoSBlock: Block already marked as ValidateFailed")
-		}
-		if blockNode.IsValidated() {
-			return blockNode, nil
-		}
+	if exists && (blockNode.IsValidateFailed() || blockNode.IsValidated()) {
+		return blockNode, nil
 	}
-	// Make sure its parent is validated. If not, we need to validate it first by calling processBlockPoS.
-	// Note that if any other ancestor is not validated, calling processBlockPoS on the parent will
-	// end up calling processBlockPoS on all of its ancestors that aren't validated yet.
-	parentBlockNode, exists := bc.blockIndex[*block.Header.PrevBlockHash]
-	if !exists {
-		// We should never really hit this, but if we do, we upsert the block w/ StatusBlockStored
-		// but not StatusBlockValidateFailed. This makes sure this function is idempotent.
-		blockNode, err = bc.storeBlockInBlockIndex(block)
-		if err != nil {
-			return blockNode, errors.Wrap(err, "isValidPoSBlock: Problem adding block to block index: ")
-		}
-		return blockNode, errors.New("isValidPoSBlock: Parent block not found in block index. This should never happen.")
+
+	// Run the validation for the parent and update the block index with the parent's status. We first
+	// check if the parent has a cached status. If so, we use the cached status. Otherwise, we run
+	// the full validation algorithm on it, then index and an use the result.
+	parentBlockNode, err := bc.validateCachedBlockPoS(block.Header.PrevBlockHash)
+	if err != nil {
```

Comment:
If there's an error, it means the validation algorithm failed. In that case, it means things are running abnormally or in some unexpected state. I'd bias towards not storing the block because to minimize the side effects from this abnormal execution of the validation algo.

That's not a strongly held opinion though

Reply:
Seems fine either way to me. Basically the caller of ProcessBlockPoS will need to request that we try to process this block again at a future time.